### PR TITLE
Serialize the Interactivity API's store from PHP and hydrate it on the client

### PR DESCRIPTION
**Pull request by DAreRodz**
## What

Ports latest changes in https://github.com/WordPress/block-hydration-experiments to add store serialization.

Changes included in this PR
- Renamed client-side "transition" to "navigation"
- Renamed `wpx` to `store`
- Fixed `class` directive when no `class` attribute is present
- Added store hydration on the browser
- Added server-side processing of directives, including store serialization

Tracking issue: https://github.com/woocommerce/woocommerce/issues/42486

## Why

To keep the Interactivity API in WooCommerce blocks up to date.

## How

For the JS part, just copy-pasting the code and updating the values in the `constants.js` file.

For the PHP part (after copy-pasting):
- Replacing `wp_` prefixes in global functions with `woo_` (excluding WP's ones, like`wp_register_script`)
- Renaming `WP_` classes to `Woo_` ones (excluding WP's ones, like `WP_HTML_Tag_Processor`)
- Renaming class files from `class-wp-` to `class-woo-`
- Locating `wp-` prefixes in directives and replacing them with `data-woo-` (except `wp-html`, in a require declaration)
- Fixing paths for required files (the file structure is different here)

For all files, removing all TODO comments using a regexp (`([^\n]+)?// TODO.*\n`).

* [x] Do not include in the Testing Notes

### CODE REVIEW

**Review by luisherranz — COMMENTED** [review]
The JS part looks good to me.

For the PHP part, maybe @ockham could take a look?

**Comment by DAreRodz** [comment]
Thanks for taking a look, @luisherranz.

BTW, for anyone following this PR: sorry for the noise with the TODO messages; I'll try not to push any of them the next time. 🙇

**Comment by DAreRodz** [comment]
@ockham, could you take a look at the PHP files, just to be sure I ported everything correctly from our repo? Thanks! 🙂

File: src/Interactivity/directives/utils.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+require_once  __DIR__ . '/class-woo-directive-store.php';
+
+function store( $data ) {
```
Comment (ockham) [inline src/Interactivity/directives/utils.php]: The function names in this file might later collide with the ones we're using in the Block Interactivity API. Have you considered prefixing them?

File: src/Interactivity/directives/woo-process-directives.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+				call_user_func( $attribute_directives[ $attribute ], $tags, $context );
+			}
+		}
+	}
+
+	return $tags;
+}
+
+// See e.g. https://github.com/WordPress/gutenberg/pull/47573.
+function is_html_void_element( $tag_name ) {
```
Comment (ockham) [inline src/Interactivity/directives/woo-process-directives.php]: Another function name to potentially prefix 😊

File: src/Interactivity/woo-directives.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
+<?php
+include_once __DIR__ . '/../../../gutenberg/lib/experimental/html/wp-html.php';
```
Comment (ockham) [inline src/Interactivity/woo-directives.php]: Note that the files in `lib/experimental/html` were recently moved to a different location. You might want to carry over the changes from https://github.com/WordPress/block-hydration-experiments/pull/162 😅

File: src/Interactivity/woo-directives.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+function woo_directives_mark_interactive_blocks( $block_content, $block, $instance ) {
+	if ( woo_directives_get_client_side_navigation() ) {
+		return $block_content;
+	}
+
+	// Append the `data-woo-ignore` attribute for inner blocks of interactive blocks.
+	if ( isset( $instance->parsed_block['isolated'] ) ) {
+		$w = new WP_HTML_Tag_Processor( $block_content );
+		$w->next_tag();
+		$w->set_attribute( 'data-woo-ignore', '' );
```
Comment (ockham) [inline src/Interactivity/woo-directives.php]: I think the preferred way of doing this is

```suggestion
		$w->set_attribute( 'data-woo-ignore', true );
```

The HTML Tag Processor is smart enough to translate that to

```html
<some-tag data-woo-ignore>
```

(and not `<some-tag data-woo-ignore="true">`, if that was your concern 😄 ).

File: src/Interactivity/woo-directives.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+	// Return if it's not interactive.
+	if ( ! block_has_support( $instance->block_type, array( 'interactivity' ) ) ) {
+		return $block_content;
+	}
+
+	// Add the `data-woo-island` attribute if it's interactive.
+	$w = new WP_HTML_Tag_Processor( $block_content );
+	$w->next_tag();
+	$w->set_attribute( 'data-woo-island', '' );
```
Comment (ockham) [inline src/Interactivity/woo-directives.php]: Same as https://github.com/woocommerce/woocommerce-blocks/pull/8447/files#r1118980149

**Comment by ockham** [comment]
> @ockham, could you take a look at the PHP files, just to be sure I ported everything correctly from our repo? Thanks! 🙂

I left a few notes! (Looking good overall though 😊)

File: src/Interactivity/woo-directives.php (hunk 4 above)
Comment (DAreRodz) [reply src/Interactivity/woo-directives.php]: Ah, nice! In that case, I think we'd have to change that in our repo as well (I just copied the code from there).

**Comment by DAreRodz** [comment]
Thanks for your reviews, @luisherranz and @ockham. I will merge this PR even though we don't have explicit approval, as I know @gigitux also reviewed the code. 😁

**Pull request merged**